base/rational.jl

```diff
@@ -145,6 +145,14 @@ function (::Type{T})(x::Rational{S}) where T<:AbstractFloat where S
     P = promote_type(T,S)
     convert(T, convert(P,x.num)/convert(P,x.den))::T
 end
+ # avoid spurious overflow (#52394).  (Needed for UInt16 or larger;
+ # we also include Int16 for consistency of accuracy.)
+Float16(x::Rational{<:Union{Int16,Int32,Int64,UInt16,UInt32,UInt64}}) =
+    Float16(Float32(x))
+Float16(x::Rational{<:Union{Int128,UInt128}}) =
+    Float16(Float64(x)) # UInt128 overflows Float32, include Int128 for consistency
+Float32(x::Rational{<:Union{Int128,UInt128}}) =
+    Float32(Float64(x)) # UInt128 overflows Float32, include Int128 for consistency
```

Comment:
Wouldn't these suffer from double-rounding accuracy loss? @oscardssmith

Reply:
they would. Doing this properly is fairly hard though.

Reply:
Also worth noting that `Float16(num)/Float16(den)` also potentially has triple rounding.

Reply:
Question is: is this problem worse than the problem this PR fixes? I.e. should we merge and improve this later, or should this wait until this PR does it "right" (whatever that means?)

Reply:
we should merge and improve later.

Reply:
That was my thought as well, but want to confirm.

Reply:
Would doing it "properly" be something like:
```julia
a = Float64(x)
b = Float32(a)
# check if we double-rounded in the wrong direction
if x > a && b < a && nextfloat(b) < a
    b = nextfloat(b)
elseif x < a && b > a && prevfloat(b) > a
    b = prevfloat(b)
end
return b
```

Reply:
A example of a test case where double rounding gives a different result is:
```jl
julia> r = 12928845309018111//18014398509481984
12928845309018111//18014398509481984

julia> Float32(r) == Float32(Float64(r))
false
```